lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,384 @@
+   */
+  ICUTransformCharFilter(Reader in, Transliterator transform, int maxRollbackBufferCapacityHint) {
+    super(in);
+    this.transform = ICUTransformFilter.optimizeForCommonCase(transform);
+    if (maxRollbackBufferCapacityHint < 0) {
+      throw new IllegalArgumentException("It is illegal to request negative rollback buffer max capacity");
+    } else if (maxRollbackBufferCapacityHint >= HARD_MAX_ROLLBACK_BUFFER_CAPACITY) {
+      // arg is positive, so user wants the largest possible buffer capacity limit
+      // we know what that is (static), protecting for overflow.
+      this.maxRollbackBufferCapacity = HARD_MAX_ROLLBACK_BUFFER_CAPACITY;
+      this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+    } else {
+      // greatest power of 2 (excluding "1") less than or equal to the specified hint
+      this.maxRollbackBufferCapacity = Integer.highestOneBit(maxRollbackBufferCapacityHint - 1) << 1;
+      if (this.maxRollbackBufferCapacity == 0) {
+        this.rollbackBuffer = null;
+      } else {
+        this.rollbackBuffer = new char[DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY];
+      }
+    }
+  }
+
+  /**
+   * Reads characters into a portion of an array. This method will block until some input is available, an I/O error
+   * occurs, or the end of the stream is reached.
+   *
+   * @param cbuf
+   *          Destination buffer
+   * @param off
+   *          Offset at which to start storing characters
+   * @param len
+   *          Maximum number of characters to read
+   * @return The number of characters read, or -1 if the end of the stream has been reached
+   * @throws IOException
+   *           If an I/O error occurs
+   */
+  @Override
+  public int read(char[] cbuf, int off, int len) throws IOException {
+    if (off < 0) throw new IndexOutOfBoundsException("specified negative array offset");
+    if (off >= cbuf.length) throw new IndexOutOfBoundsException("specified offset exceeds buffer length");
+    if (len <= 0) throw new IndexOutOfBoundsException("non-positive length specified");
+    if (len > cbuf.length - off) throw new IndexOutOfBoundsException("requested end array index exceeds buffer length");
+
+    // !inputFinished || output remains to be flushed
+    while (!inputFinished || position.start > outputCursor) {
+      if (position.start > outputCursor) {
+        return outputFromResultBuffer(cbuf, off, len);
+      }
+
+      int resLen = readFromIoNormalizeUptoBoundary();
+      if (resLen > 0) {
+        return outputFromResultBuffer(cbuf, off, len);
+      }
+
+      if (!readInputToBuffer()) {
+        final int preStart = position.start;
+        final int preLimit;
+        final int bufferLength = buffer.length();
+        cursorUpdate:
+        {
+          if (preStart < bufferLength) {
+            // if last char is a lead surrogate, transform won't handle it properly anyway
```

Comment:
Yes, that's the case. I agree with the approach you outline. I'll double-check, but I think this code doesn't try to _correct_ the error (i.e., doesn't trim the invalid input off the end) -- it just doesn't attempt to transform it. So the approach is basically "I don't care what you gave me, I'll do what I can with it and pass it on" -- making the transliterator totally unopinionated, and not treating it as a validator. Garbage in, garbage out, no judgment :-)
I could see throwing an error instead if you think that'd be preferable.

Reply:
I think we should provide whatever behavior the Transliterator would if you gave *it* malformed input as a String, and ideally we'd do that without any special error-handling code of our own, just for simplicity's sake, and in order to track Transliterator's implementation should it change.

Reply:
IIRC I actually did encounter this situation, and the ICU Transliterator didn't throw an error itself, it just screwed up the positions in an insidious way that eventually manifested as illegal offsets.
Ultimately though there's nothing in the ICU abstract `Transliterator` class that indicates how `finishTransliterate(...)` behaves on malformed input (whether it throws errors, recovers, whatever); it just delegates to the abstract `handleTransliterate(...)` method, so perhaps the best thing to do would be to proactively throw an error -- that's also the _easiest_ thing to do since the current code goes to the trouble of detecting this situation anyway.